### PR TITLE
[Mi2]:Reduce the total number of Break and Continue Statements

### PR DESCRIPTION
**Pull request by NarraVivek**
**What's in the PR**
Reduction of the total number of Break and Continue statements in a loop or method is performed.

Note: few instances which had empty conditions were studied and left out
without removing the continue statements. Break statements which were
detected in the code smells are very few and are required to be present for
their respective cases.

### CODE REVIEW

File: inception-imls-opennlp/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/opennlp/ner/OpenNlpNerRecommender.java [hunk 1]
```diff
@@ -352,7 +352,7 @@ private NameSample createNameSample(CAS aCas, AnnotationFS aSentence,
             if (beginToken == null || endToken == null) {
                 LOG.warn("Skipping annotation not starting/ending at token boundaries: [{}-{}, {}]",
                         annotation.getBegin(), annotation.getEnd(), label);
-                continue;
```
Comment (munterkalmsteiner) [inline inception-imls-opennlp/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/opennlp/ner/OpenNlpNerRecommender.java:355]: No, the `continue` is needed here! Without it, the loop continues to execute which is not wanted. If `beginToken` or `endToken` is null, the loop should continue with the next iteration.

File: inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/reification/NoReification.java [hunk 2]
```diff
@@ -112,12 +112,12 @@
             if (value == null) {
                 // Can this really happen?
                 log.warn("Property with null value detected.");
-                continue;
```
Comment (munterkalmsteiner) [inline inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/reification/NoReification.java]: Same here: removing `continue` introduces a bug in the code.

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/evaluation/SimulationLearningCurvePanel.java [hunk 3]
```diff
@@ -251,7 +251,7 @@ private LearningCurve renderChart()
                 
                 if (evaluationResult.isEvaluationSkipped()) {
                     LOG.warn("Evaluation skipped. Chart cannot to be shown");
-                    continue;
```
Comment (munterkalmsteiner) [inline inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/evaluation/SimulationLearningCurvePanel.java:254]: Not correct. Changes behavior.

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java [hunk 4]
```diff
@@ -850,7 +850,7 @@ public Predictions computePredictions(User aUser, Project aProject,
                 if (recommenders.isEmpty()) {
                     log.trace("[{}]: No active recommenders on layer [{}]", username,
                             layer.getUiName());
-                    continue;
```
Comment (munterkalmsteiner) [inline inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java]: Probably also not correct. Changes behavior

**Review by munterkalmsteiner — CHANGES_REQUESTED** [review]
Hm, I think most of the changes you propose (removing `continue`) actually changes the behavior of the code (actually, introduces unwanted behavior, i.e. bugs). I think the code smell is, in some cases, correct. However, your fix is not.

**Review by NarraVivek — COMMENTED** [review]
I have reduced the statements which affect the program

**Review by munterkalmsteiner — CHANGES_REQUESTED** [review]
You should have looked at the other removals of `continue` too... they are all not correct. I think all of the changes introduce a bug.
The solution is a different refactoring, not to simply remove `continue`, since this changes the behavior of the code.
I think it is difficult to propose a better (in terms of code maintainability) solution than using `continue`. Please investigate, provide a solution (refactor all or only some of the finding instances) OR motivate why it is not possible to improve the code.

**Comment by NarraVivek** [comment]
Sir,
I have investigated the various way of handling the continue and break statements, but have found that the usage of these statements is a bad practice in the first place and is often used when there is no other option while coding which is observed in this application. There might be an alternative in changing the control flow of the code which might lead to more issues or change the behavior of the code. I think that use of these statements was maybe necessary to the developers which is resulting the new bugs upon their removal. 

If needed I will add the statements to their original locations. Please let me know if there is anything wrong with my insight.

**Comment by munterkalmsteiner** [comment]
Hi @NarraVivek. Please report here the different ways how to refactor `continue` and `break` statements (deleting them is not a correct way). 
You just need to explain the different approaches here, no need to actually implement them in the code if it makes the code more complicated (that would be against the idea of refactoring). However, I (and Deepika probably) want to make sure you understand the issue well.

**Comment by NarraVivek** [comment]
I understand now, Apologies for the confusion.
With regard to this particular issue I think there are few ways to avoid the usage of "continue" and "break". As mentioned above in the previous comment, usage of these statements is a bad practice and are least preferred while programming. Although the use of "break" statements is only seen in switch cases which is mandatory or required to be placed as the control flow needs to end after an action. The "continue" statement can be avoided or refactored by using the "goto" or "return" functionality after making significant changes to the program code which is basically providing the reference of the next function to be performed or transfer the control from one area to another in the code.

**Comment by munterkalmsteiner** [comment]
@NarraVivek, and why do these strategies not work with the code you have tried to refactor?

**Comment by NarraVivek** [comment]
@munterkalmsteiner These strategies could be used if I had more experience and time in using this application and know where the exact functions calls and utilization's were made. We can observe that most of the instances are in the form of "for" loops, I believe any attempt of mine to implement these alternatives will change the control and execution flow of the application if misplaced.

**Comment by NarraVivek** [comment]
And another factor which we have not accounted is : to change or specify any control flow changes, I think it will be crucial to adjust the code structure. It might pose a risk if its required to do so.

**Comment by munterkalmsteiner** [comment]
@NarraVivek no experience and no time are no excuses ;) You are here to learn something, after all. If there is a risk that the code changes might affect the behavior of the program, the first step would be to write test cases to make sure changes do *not* affect program behavior. 

I'll give you now an example of what I would have expected as an answer.

Looking the the first change you proposed:

```
while (SOME_CONDITION) {
    if (beginToken == null || endToken == null) {
        continue;
    }
    DO_SOMETHING_IMPORTANT_HERE
}
```

This could be rewritten as:

```
while (SOME_CONDITION) {
    if (beginToken != null && endToken != null) {
        DO_SOMETHING_IMPORTANT_HERE
    }
}
```

The two code snippets do exactly the same. The problem with the second one is that DO_SOMETHING_IMPORTANT_HERE is nested within an IF statement. Not a huge problem if there is only 1 level of nesting, but it could be more levels and then the important code is deep down in if statements and it is difficult to follow what is going on. Hence, the solution with continue which is usually done at the beginning of the loop to indicate early under what condition the loop should be jump over one iteration.
Hope this clarifies the code smell for you. Not all smells are true positives, there is also a grey area where one can argue both for and against a smell.

**Comment by NarraVivek** [comment]
Yes, I can see that I was unable to illustrate this idea through an example. I apologise for that. It is clear and known now about this issue. It was a conflict of interest among our group also whether this refactoring was worth changing when we were working under this PR.

**Pull request closed without merging**